File: backend2/models/ExchangeTransaction.js
module.exports = (sequelize, DataTypes) => {
  const ExchangeTransaction = sequelize.define('ExchangeTransaction', {
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true
    },
    user_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      references: {
        model: 'users',
        key: 'id'
      }
    },
    reference_number: {
      type: DataTypes.STRING(255),
      allowNull: true,
      unique: true
    },
    from_currency: {
      type: DataTypes.STRING(3),
      allowNull: false,
      validate: {
        len: [3, 3],
        isAlpha: true,
        isUppercase: true
      }
    },
    to_currency: {
      type: DataTypes.STRING(3),
      allowNull: false,
      validate: {
        len: [3, 3],
        isAlpha: true,
        isUppercase: true
      }
    },
    from_amount: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: false,
      validate: {
        min: 0
      }
    },
    to_amount: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: false,
      validate: {
        min: 0
      }
    },
    exchange_rate: {
      type: DataTypes.DECIMAL(15, 6),
      allowNull: false,
      validate: {
        min: 0
      }
    },
    fee_amount: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: true,
      validate: {
        min: 0
      }
    },
    status: {
      type: DataTypes.ENUM('PENDING', 'PROCESSING', 'COMPLETED', 'FAILED', 'CANCELLED'),
      allowNull: false,
      defaultValue: 'PENDING'
    },
    completed_at: {
      type: DataTypes.DATE,
      allowNull: true
    }
  }, {
    tableName: 'exchange_transactions',
    timestamps: true,
    underscored: true,
    hooks: {
      beforeCreate: (transaction) => {
        if (!transaction.reference_number) {
          transaction.reference_number = `TXN-${Date.now()}-${Math.floor(Math.random() * 1000)}`;
        }
      }
    },
    indexes: [
      {
        fields: ['user_id']
      },
      {
        unique: true,
        fields: ['reference_number']
      },
      {
        fields: ['status']
      },
      {
        fields: ['created_at']
      },
      {
        fields: ['from_currency', 'to_currency']
      }
    ]
  });

  ExchangeTransaction.associate = function(models) {
    ExchangeTransaction.belongsTo(models.User, {
      foreignKey: 'user_id',
      as: 'user'
    });
  };

  return ExchangeTransaction;
}; 

File: backend/src/main/java/com/currencyexchange/entity/ExchangeTransaction.java
package com.currencyexchange.entity;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "exchange_transactions")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ExchangeTransaction {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;
    
    @Column(name = "from_currency", nullable = false, length = 3)
    private String fromCurrency;
    
    @Column(name = "to_currency", nullable = false, length = 3)
    private String toCurrency;
    
    @Column(name = "from_amount", nullable = false, precision = 15, scale = 2)
    private BigDecimal fromAmount;
    
    @Column(name = "to_amount", nullable = false, precision = 15, scale = 2)
    private BigDecimal toAmount;
    
    @Column(name = "exchange_rate", nullable = false, precision = 15, scale = 6)
    private BigDecimal exchangeRate;
    
    @Column(name = "fee_amount", precision = 15, scale = 2)
    private BigDecimal feeAmount;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionStatus status;
    
    @Column(name = "reference_number", unique = true)
    private String referenceNumber;
    
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "completed_at")
    private LocalDateTime completedAt;
    
    public enum TransactionStatus {
        PENDING,
        PROCESSING,
        COMPLETED,
        FAILED,
        CANCELLED
    }
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (referenceNumber == null) {
            referenceNumber = generateReferenceNumber();
        }
    }
    
    private String generateReferenceNumber() {
        return "TXN-" + System.currentTimeMillis() + "-" + (int)(Math.random() * 1000);
    }
} 